types/react-autosuggest/index.d.ts

```diff
@@ -5,84 +5,200 @@
+        { [P in U]: never } & { [x: string]: never })[T];
+
+    /** @internal */
+    type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>;
+
+    interface SuggestionsFetchRequestedParams {
+        value: string;
+        reason:
+            | 'input-changed'
+            | 'input-focused'
+            | 'escape-pressed'
+            | 'suggestions-revealed'
+            | 'suggestion-selected';
+    }
+
+    interface RenderSuggestionParams {
+        query: string;
+        isHighlighted: boolean;
+    }
+
+    interface SuggestionHighlightedParams {
+        suggestion: any;
+    }
+
+    interface ChangeEvent {
+        newValue: string;
+        method: 'down' | 'up' | 'escape' | 'enter' | 'click' | 'type';
+    }
+
+    interface BlurEvent<TSuggestion> {
+        highlightedSuggestion: TSuggestion;
+    }
+
+    interface InputProps<TSuggestion>
+        extends Omit<React.InputHTMLAttributes<any>, 'onChange' | 'onBlur'> {
+        onChange(event: React.FormEvent<any>, params?: ChangeEvent): void;
+        onBlur?(event: React.FormEvent<any>, params?: BlurEvent<TSuggestion>): void;
+        value: string;
+        [key: string]: any;
+    }
+
+    interface SuggestionSelectedEventData<TSuggestion> {
+        suggestion: TSuggestion;
+        suggestionValue: string;
+        suggestionIndex: number;
+        sectionIndex: number | null;
+        method: 'click' | 'enter';
+    }
+
+    type ThemeKey =
+        | 'container'
+        | 'containerOpen'
+        | 'input'
+        | 'inputOpen'
+        | 'inputFocused'
+        | 'suggestionsContainer'
+        | 'suggestionsContainerOpen'
+        | 'suggestionsList'
+        | 'suggestion'
+        | 'suggestionFirst'
+        | 'suggestionHighlighted'
+        | 'sectionContainer'
+        | 'sectionContainerFirst'
+        | 'sectionTitle';
+
+    type Theme =
+        | Record<string, string | React.CSSProperties>
+        | Partial<Record<ThemeKey, string | React.CSSProperties>>;
+
+    interface RenderSuggestionsContainerParams {
+        containerProps: {
+            id: string;
+            key: string;
+            ref: any;
+            style: any;
+        };
+        children: React.ReactNode;
+        query: string;
+    }
+
+    // types for functions - allowing reuse externally - e.g. as props and bound in the constructor
+    type GetSectionSuggestions<TSuggestion> = (section: any) => TSuggestion[];
+    type GetSuggestionValue<TSuggestion> = (suggestion: TSuggestion) => string;
+    type OnSuggestionHighlighted = (params: SuggestionHighlightedParams) => void;
+    type SuggestionsFetchRequested = (request: SuggestionsFetchRequestedParams) => void;
+    type OnSuggestionsClearRequested = () => void;
+    type OnSuggestionSelected<TSuggestion> = (
+        event: React.FormEvent<any>,
+        data: SuggestionSelectedEventData<TSuggestion>,
+    ) => void;
+    type RenderInputComponent<TSuggestion> = (inputProps: InputProps<TSuggestion>) => JSX.Element;
```

Comment:
I would suggest it will be more correct to use `React.ReactNode` instead of `JSX.Element`.